src/toxcore/dht/server.rs

```diff
@@ -170,26 +179,29 @@ impl Server {
         match packet {
             DhtPacket::PingRequest(packet) => {
                 debug!("Received ping request");
-                let client = self.create_client(&addr, packet.pk);
-                self.handle_ping_req(client, packet)
+                self.handle_ping_req(packet, addr)
             },
             DhtPacket::PingResponse(packet) => {
                 debug!("Received ping response");
-                if let Some(client) = self.get_client(&packet.pk) {
-                    self.handle_ping_resp(client, packet)
+                let mut state = self.state.write();
```

Comment:
You must call `handle_*(packet, addr)` here, while you take locks inside handle_* function.